### Importing libraries

In [22]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import sklearn
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.metrics import classification_report

from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import MultinomialNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier

from sklearn.externals import joblib

In [190]:
data = pd.read_csv("Data/reddit-top-flairs-cleaned.csv")

In [191]:
data.head()

id                                                url  \
0  g2ct57  https://www.reddit.com/r/india/comments/g2ct57...   
1  futac9  https://www.reddit.com/r/india/comments/futac9...   
2  ff8sth                https://i.redd.it/yjo9wpy38el41.jpg   
3  fpaj1w  https://theprint.in/india/hit-by-backlash-over...   
4  fxs1vy  https://www.timesnownews.com/india/article/pol...   

                                               title  \
0                             polite request indians   
1  pitting community political party fucking stupid    
2  new political party gave full front page ad po...   
3  hit backlash posts lack medical gear doctors g...   
4  politics time corona wb cm questions centre s ...   

                                                body     flair  \
0  nt know situation countries india seen lot org...  Politics   
1  first let start saying stupid whatever muslims...  Politics   
2                                                nil  Politics   
3                                                nil  Politics   
4                                                nil  Politics   

                                         dirty_title  \
0               A polite request to all Indians here   
1  Pitting a community against a political party ...   
2  A new political party gave a full front page a...   
3  Hit by backlash over posts on lack of medical ...   
4  Politics in the time of corona: WB CM question...   

                                          dirty_body  num_words_title  \
0  I don't know if it is the same situation in ot...                3   
1  First of all let me start by saying it was stu...                6   
2                                                NIL               12   
3                                                NIL               26   
4                                                NIL               15   

   num_words_body  num_unique_words_title  num_unique_words_body  \
0             109                       3                     83   
1             157                       6                    119   
2               1                      12                      1   
3               1                      23                      1   
4               1                      15                      1   

   num_chars_title  num_chars_body  
0               22             757  
1               49            1067  
2               73               3  
3              191               3  
4               84               3

### Dropping unwanted columns and shuffling the data.

In [192]:
data.drop(['dirty_title','dirty_body','num_words_body','num_unique_words_title'
           ,'num_unique_words_body','num_chars_title','num_words_title','num_chars_body','url','id'], axis=1)

data.sample(frac=1)

id                                                url  \
366  fk25rj  https://www.reddit.com/r/india/comments/fk25rj...   
670  cjnx1s  https://www.reddit.com/r/india/comments/cjnx1s...   
563  aoef6r                https://i.redd.it/p491fl35yaf21.jpg   
30   g0y7zg  https://www.reddit.com/r/india/comments/g0y7zg...   
49   g3jfil  https://in.news.yahoo.com/indian-tricolour-pro...   
..      ...                                                ...   
497  d76uvt  https://www.reddit.com/r/india/comments/d76uvt...   
32   f1a6p2  https://theprint.in/opinion/pov/talking-to-cab...   
40   g1nd9a  https://theprint.in/india/governance/ias-edge-...   
833  7yz12m  https://www.reddit.com/r/india/comments/7yz12m...   
150  7lu19z                https://i.redd.it/c15yam3gst501.jpg   

                                                 title  \
366   askindia government put emi payment hold apri...   
670  query data science courses institutions detail...   
563  shimla last night s snowfall mobile mi a2 phot...   
30   random daily discussion thread april 14 2020 0...   
49   indian tricolour projected onto switzerland s ...   
..                                                 ...   
497  economic shit storm incoming brace impact fail...   
32   talking cab drivers politics oldest cliché ube...   
40   ias edge proves winner chief secretaries overs...   
833                        need help depression anger    
150              nonpolitical company s ad pic middle    

                                                  body               flair  \
366  advent corona virus india money flow severely ...            AskIndia   
670  hello folks friend query regarding data scienc...    Business/Finance   
563                                                nil         Photography   
30   beep boop bot i am a bot if any problem askawa...            Politics   
49                                                 nil            Politics   
..                                                 ...                 ...   
497  getting bombarded let say scenario rather pess...      Policy/Economy   
32                                                 nil            Politics   
40                                                 nil            Politics   
833  recently got heart broken realised unrequited ...  Science/Technology   
150                                                nil       Non-Political   

                                           dirty_title  \
366  [AskIndia] Why has the government not put all ...   
670  A query about Data Science: Courses, Instituti...   
563  Shimla after last night's snowfall. Mobile - M...   
30   Random Daily Discussion Thread - April 14, 202...   
49   Indian Tricolour Projected onto Switzerland's ...   
..                                                 ...   
497  Economic Shit Storm Incoming brace for impact ...   
32   Talking to cab drivers about politics is the o...   
40   IAS edge proves a winner as chief secretaries ...   
833               Need help with depression and anger.   
150  [Non-Political] Which company's ad is in the p...   

                                            dirty_body  num_words_title  \
366  With the advent of Corona virus in India, mone...                8   
670  Hello folks, a friend had a query regarding Da...                8   
563                                                NIL                9   
30   ^Beep ^Boop ^Bot, ^I ^am ^a ^bot! ^if ^any ^pr...                8   
49                                                 NIL               11   
..                                                 ...              ...   
497  Before getting bombarded, let me say that it i...               12   
32                                                 NIL                9   
40                                                 NIL               11   
833  Recently got my heart broken, realised it was ...                4   
150                                                NIL      

In [131]:
top_flairs = ["Politics", "Non-Political", "Coronavirus", "AskIndia", "Policy/Economy", "[R]eddiquette", 
              "Photography", "Business/Finance", "Sports", "Science/Technology"]

### The columns of "title" and "body" are combined as one feature as "input_features" and "cat" stores the output label. Train test split is also done.

In [132]:
input_features = data["title"] + " "+ data["body"]
data = data.assign(input_features = input_features)

y = data.flair

x_train, x_test, y_train, y_test = train_test_split(input_features,y, test_size=0.3)

print("x_train dim:",x_train.shape, "\ty_train dim:", y_train.shape)
print("x_test dim:",x_test.shape, "\ty_test dim:", y_test.shape)

x_train dim: (692,) 	y_train dim: (692,)
x_test dim: (297,) 	y_test dim: (297,)


## Logistic Regression

It is a typical classifier and most pervasively used one too. It has great interpretability properties.
### Parameters:
1. C = inverse regularization parameter; C = 1/λ;  It’s a penalty term meant regulate against overfitting.
2. max_iter = maximum iterations done.

In [138]:
# Logistic Regression
logistic = Pipeline([('cv', CountVectorizer()),('tfidf', TfidfTransformer()),('clf', LogisticRegression(C=1000, max_iter=1000))])
logistic.fit(x_train, y_train)

y_pred = logistic.predict(x_test)

accuracy = (accuracy_score(y_pred, y_test)*100)
print("Accuracy: %.2f" % accuracy +'%')
print(classification_report(y_test, y_pred,target_names=top_flairs))

Accuracy: 77.10%
                    precision    recall  f1-score   support

          Politics       0.79      0.82      0.81        33
     Non-Political       0.55      0.55      0.55        29
       Coronavirus       0.93      0.97      0.95        29
          AskIndia       0.85      0.82      0.84        34
    Policy/Economy       0.82      0.97      0.89        29
     [R]eddiquette       0.65      0.78      0.71        36
       Photography       0.73      0.73      0.73        30
  Business/Finance       0.96      0.79      0.87        29
            Sports       0.74      0.81      0.77        21
Science/Technology       0.71      0.44      0.55        27

          accuracy                           0.77       297
         macro avg       0.77      0.77      0.77       297
      weighted avg       0.77      0.77      0.77       297



/Users/keshav/anaconda3/envs/reddit/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


### Inference from Logistic Regression:
1. The model has a total accuracy of **77%**, which is good.
2. The most performing flair category was **"Coronavirus"**.
3. The least performing flair categories were **"Non-Political"** and **"Sciene/Technology"**.

## NAIVE BAYES CLASSIFIER
Using Naive Bayes text classifiers might be a really good idea initially, especially if there’s not much training data available and computational resources are scarce. One of the most suitable variants for text is the multinomial variant.

In [133]:
# Naive Bayes
naive = Pipeline([('cv', CountVectorizer()),('tfidf', TfidfTransformer()),('nb', MultinomialNB())])
naive.fit(x_train, y_train)
y_pred = naive.predict(x_test)

accuracy = (accuracy_score(y_pred, y_test)*100)
print("Accuracy: %.2f" % accuracy +'%')
print(classification_report(y_test, y_pred,target_names=top_flairs))

Accuracy: 67.68%
                    precision    recall  f1-score   support

          Politics       0.59      0.73      0.65        33
     Non-Political       0.37      0.69      0.48        29
       Coronavirus       0.88      0.97      0.92        29
          AskIndia       0.89      0.74      0.81        34
    Policy/Economy       0.82      0.93      0.87        29
     [R]eddiquette       0.59      0.61      0.60        36
       Photography       0.74      0.67      0.70        30
  Business/Finance       0.89      0.59      0.71        29
            Sports       0.67      0.76      0.71        21
Science/Technology       1.00      0.07      0.14        27

          accuracy                           0.68       297
         macro avg       0.74      0.67      0.66       297
      weighted avg       0.74      0.68      0.66       297



### Inference from Naive Bayes Classifier:
1. The model has a total accuracy of **68%**, which is decent.
2. The most performing flair category was **"Coronavirus"**.
3. The least performing flair category was **"Sciene/Technology"**.

## Random Forest Classifier

Random forest classifier suits most multi-class classification problem, also they have good interpretability and work faster.
### Parameters:
1. n_estimators = the number of trees inthe forest.

In [144]:
# Random Forest
random = Pipeline([('cv', CountVectorizer()),('tfidf', TfidfTransformer()),('rf', RandomForestClassifier(n_estimators = 1000))])
random.fit(x_train, y_train)

y_pred = random.predict(x_test)

accuracy = (accuracy_score(y_pred, y_test)*100)
print("Accuracy: %.2f" % accuracy +'%')
print(classification_report(y_test, y_pred,target_names=top_flairs))

Accuracy: 82.15%
                    precision    recall  f1-score   support

          Politics       0.69      0.82      0.75        33
     Non-Political       0.73      0.55      0.63        29
       Coronavirus       0.94      1.00      0.97        29
          AskIndia       0.94      0.91      0.93        34
    Policy/Economy       0.93      0.97      0.95        29
     [R]eddiquette       0.83      0.83      0.83        36
       Photography       0.76      0.83      0.79        30
  Business/Finance       0.96      0.90      0.93        29
            Sports       0.71      0.81      0.76        21
Science/Technology       0.68      0.56      0.61        27

          accuracy                           0.82       297
         macro avg       0.82      0.82      0.81       297
      weighted avg       0.82      0.82      0.82       297



In [28]:
joblib.dump(random, 'random-forest.pkl')

['random-forest.pkl']

### Inference from Random Forest Classifier:
1. The model has a total accuracy of **82%**, which is goood.
2. The most performing flair category was **"Coronavirus"**.
3. The least performing flair category was **"Sciene/Technology"**.

## k- Nearest Neighbours Classifier
### Parameters:
1. n_neighbours = number of neighbours to take into consideration.

In [154]:
# k-Nearest Neighbours
neighbours = Pipeline([('vect', CountVectorizer()),('tfidf', TfidfTransformer()),('clf', KNeighborsClassifier(n_neighbors=5))])
neighbours.fit(x_train, y_train)
y_pred = neighbours.predict(x_test)

accuracy = (accuracy_score(y_pred, y_test)*100)
print("Accuracy: %.2f" % accuracy +'%')
print(classification_report(y_test, y_pred,target_names=top_flairs))

Accuracy: 61.62%
                    precision    recall  f1-score   support

          Politics       0.43      0.45      0.44        33
     Non-Political       0.36      0.59      0.45        29
       Coronavirus       0.68      0.93      0.78        29
          AskIndia       0.73      0.79      0.76        34
    Policy/Economy       0.84      0.90      0.87        29
     [R]eddiquette       0.56      0.64      0.60        36
       Photography       0.68      0.43      0.53        30
  Business/Finance       0.76      0.55      0.64        29
            Sports       0.79      0.52      0.63        21
Science/Technology       0.67      0.30      0.41        27

          accuracy                           0.62       297
         macro avg       0.65      0.61      0.61       297
      weighted avg       0.64      0.62      0.61       297



### Inference from k- Nearest Neighbours Classifier:
1. The model has a total accuracy of **62%**, which is goood.
2. The most performing flair category was **"Policy/Economy"**.
3. The least performing flair category was **"Sciene/Technology"**.

# OUT OF THE 4 MODELS USED, RANDOM FOREST PERFORMED THE BEST WITH A TOTAL ACCURACY OF 82%